File: app/src/main/java/com/example/avjindersinghsekhon/minimaltodo/DeleteNotificationService.java
package com.example.avjindersinghsekhon.minimaltodo;

import android.app.IntentService;
import android.content.Intent;
import android.content.SharedPreferences;

import com.example.avjindersinghsekhon.minimaltodo.contract.Contract;
import com.example.avjindersinghsekhon.minimaltodo.data.ToDoItem;
import com.example.avjindersinghsekhon.minimaltodo.data.source.todo.StoreRetrieveData;

import java.util.ArrayList;
import java.util.UUID;

public class DeleteNotificationService extends IntentService {

    private StoreRetrieveData storeRetrieveData;
    private ArrayList<ToDoItem> mToDoItems;
    private ToDoItem mItem;

    public DeleteNotificationService(){
        super("DeleteNotificationService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        storeRetrieveData = new StoreRetrieveData(this, Contract.FILENAME);
        UUID todoID = (UUID)intent.getSerializableExtra(TodoNotificationService.TODOUUID);

        mToDoItems = loadData();
        if(mToDoItems!=null){
            for(ToDoItem item : mToDoItems){
                if(item.getIdentifier().equals(todoID)){
                    mItem = item;
                    break;
                }
            }

            if(mItem!=null){
                mToDoItems.remove(mItem);
                dataChanged();
                saveData();
            }

        }

    }

    private void dataChanged(){
        SharedPreferences sharedPreferences = getSharedPreferences(Contract.SHARED_PREF_DATA_SET_CHANGED, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putBoolean(Contract.CHANGE_OCCURED, true);
        editor.apply();
    }

    private void saveData(){
        try{
            storeRetrieveData.saveToFile(mToDoItems);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        saveData();
    }

    private ArrayList<ToDoItem> loadData(){
        try{
            return storeRetrieveData.loadFromFile();
        }
        catch (Exception e){
            e.printStackTrace();
        }

        return null;

    }
}

File: app/src/main/java/com/example/avjindersinghsekhon/minimaltodo/DeleteNotificationService.kt
package com.example.avjindersinghsekhon.minimaltodo

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import org.json.JSONException
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by patternoid on 2017. 7. 21..
 */
class DeleteNotificationService : IntentService {

    private var storeRetrieveData   : StoreRetrieveData?    = null
    private var mToDoItems          : ArrayList<ToDoItem?>? = null
    private var mItem               : ToDoItem?             = null


    constructor() : super("DeleteNotificationService")



    override fun onHandleIntent(intent : Intent?) {

        val todoID: UUID    = intent?.getSerializableExtra(TodoNotificationService.TODOUUID) as UUID

        storeRetrieveData   = StoreRetrieveData(this, MainActivity.FILENAME)
        mToDoItems          = loadData()
        mItem               = mToDoItems!!.filter { item -> item?.identifier!! == todoID }.singleOrNull()

        mItem?.let {
            mToDoItems!!.remove(mItem)
            dataChanged()
            saveData()
        }
    }



    private fun dataChanged(){

        baseContext.getSharedPreferences(MainActivity.SHARED_PREF_DATA_SET_CHANGED, Context.MODE_PRIVATE).edit().run {
            putBoolean(MainActivity.CHANGE_OCCURED, true)
        }.apply()

    }



    private fun saveData(){
        try{
            storeRetrieveData?.saveToFile(mToDoItems)
        }
        catch( e : Exception ){
            e.printStackTrace()
        }
    }



    override fun onDestroy(){
        super.onDestroy()
        saveData()
    }



    private fun loadData() : ArrayList<ToDoItem?>{

        var loadData : ArrayList<ToDoItem?> = ArrayList<ToDoItem?>(0)

        try{
            loadData = storeRetrieveData!!.loadFromFile()

            if( loadData == null )
                throw KotlinNullPointerException()
        }

        catch( e : Exception){

            when(e){
                is KotlinNullPointerException,
                is JSONException,
                is IOException -> {
                    e.printStackTrace()
                }
                else -> throw e
            }
        }

        return loadData
    }
}